### PR TITLE
SDI-266 Used refactored service

### PR DESCRIPTION
**Pull request by andymarke**
Service convert to Kotlin for readability. 

Existing full integration tests remain unchanged showing everything works exactly as before.

Existing small unit test for nomis number generator ported as is

### CODE REVIEW

File: src/main/java/uk/gov/justice/hmpps/prison/api/model/PrisonerIdentifier.java [hunk 1]
```diff
@@ -8,6 +8,7 @@
 import lombok.EqualsAndHashCode;
 import lombok.NoArgsConstructor;
 import lombok.ToString;
+import lombok.With;
```
Comment (andymarke) [inline src/main/java/uk/gov/justice/hmpps/prison/api/model/PrisonerIdentifier.java:11]: Kotlin can use `with` but not builders

File: src/main/java/uk/gov/justice/hmpps/prison/service/receiveandtransfer/PrisonerCreationService.kt [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+
+    return offenderTransformer.transform(prisoner)
+  }
+
+  fun getNextPrisonerIdentifier(): PrisonerIdentifier {
+    var retries = 0
+    var updated: Boolean
+    var nextSequence: NomsIdSequence
+    var currentSequence: NomsIdSequence
+    do {
```
Comment (andymarke) [inline src/main/java/uk/gov/justice/hmpps/prison/service/receiveandtransfer/PrisonerCreationService.kt:87]: TODO - convert to Kotlin idiom

File: src/main/java/uk/gov/justice/hmpps/prison/service/receiveandtransfer/PrisonerCreationService.kt [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+    }
+    return PrisonerIdentifier().withId(currentSequence.prisonerIdentifier)
+  }
+
+  private fun gender(code: String): Result<Gender> =
+    genderRepository.findByIdOrNull(ReferenceCode.Pk(Gender.SEX, code))?.let { success(it) } ?: failure(
+      EntityNotFoundException.withMessage("Gender $code not found")
+    )
+
+  private fun ethnicity(code: String?): Result<Ethnicity>? =
```
Comment (mikehalmamoj) [inline src/main/java/uk/gov/justice/hmpps/prison/service/receiveandtransfer/PrisonerCreationService.kt:103]: I'm not sure why these are nullable? Could return type to be `Result<Ethnicity?>`?

File: src/main/java/uk/gov/justice/hmpps/prison/service/receiveandtransfer/PrisonerCreationService.kt [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+  private fun validCroNumber(croNumber: String?): Result<String>? =
+    croNumber?.takeIf { it.isNotBlank() }?.checkForDuplicateCroNumber()
+
+  private fun String.checkForDuplicateCroNumber(): Result<String> =
+    offenderIdentifierRepository.findByIdentifierTypeAndIdentifier("CRO", this)
+      .takeIf { matches -> matches.isNotEmpty() }?.let { matches ->
+        failure(matches.toCroMatchFailure())
+      } ?: success(this)
+
+  private fun validUniqueUppercaseNames(requestToCreate: RequestToCreate): Result<Pair<String, String>> {
```
Comment (mikehalmamoj) [inline src/main/java/uk/gov/justice/hmpps/prison/service/receiveandtransfer/PrisonerCreationService.kt:143]: Minor, but all these `Pair<String, String>` could be a value class for readability. Maybe called `FirstAndLastName`.

**Review by mikehalmamoj — APPROVED** [review]

**Pull request merged**